src/grisp_connect_client.erl

```diff
@@ -147,19 +106,57 @@ handle_event(state_timeout, wait, connecting, Data) ->
             ?LOG_INFO(#{event => waiting_ws_connection}),
             {keep_state_and_data, [{state_timeout, ?STD_TIMEOUT, wait}]}
     end;
-handle_event(cast, disconnected, connecting, _Data) ->
+connecting(cast, disconnected, _Data) ->
     repeat_state_and_data;
+?HANDLE_COMMON.
 
-% CONNECTED
-handle_event(enter, _OldState, connected, _Data) ->
+connected(enter, _OldState, _Data) ->
     grisp_connect_log_server:start(),
     keep_state_and_data;
-handle_event(cast, disconnected, connected, Data) ->
+connected({call, From}, is_connected, _) ->
+    {keep_state_and_data, [{reply, From, true}]};
+connected(cast, disconnected, Data) ->
     ?LOG_WARNING(#{event => disconnected}),
     grisp_connect_log_server:stop(),
     {next_state, waiting_ip, Data};
+connected(cast, {handle_message, Payload}, #data{requests = Requests} = Data) ->
+    Replies = grisp_connect_api:handle_msg(Payload),
+    % A reduce operation is needed to support jsonrpc batch comunications
+    case Replies of
+        [] ->
+            keep_state_and_data;
+        [{request, Response}] -> % Response for a GRiSP.io request
+            grisp_connect_ws:send(Response),
+            keep_state_and_data;
+        [{response, ID, Response}] -> % GRiSP.io response
+            {OtherRequests, Actions} = dispatch_response(ID, Response, Requests),
+            {keep_state, Data#data{requests = OtherRequests}, Actions}
+    end;
+connected({call, From}, {request, Method, Type, Params},
+          #data{requests = Requests} = Data) ->
+    {ID, Payload} = grisp_connect_api:request(Method, Type, Params),
+    grisp_connect_ws:send(Payload),
+    NewRequests = Requests#{ID => From},
+    {keep_state,
+     Data#data{requests = NewRequests},
+     [{{timeout, ID}, request_timeout(), request}]};
+?HANDLE_COMMON.
 
-handle_event(E, OldS, NewS, Data) ->
+% Common event handling appended as last match case to each state_function
+handle_common(cast, connect, _State, _Data)->
+    keep_state_and_data;
+handle_common({call, From}, is_connected, _, _) ->
```

Comment:
To make the code more robust against adding bugs later and for readability, I would add `when State =/= connected` here, even though it's theoretically redundant since the case `State =:= connected` is already handled in the state function. Just for readability and in case the pattern match is wrong or something in the state function `connected/3` or here.

Reply:
Ah it was there and I removed it eheh, ok I'll put it back